cpp/include/cuspatial/experimental/detail/polygon_bounding_boxes.cuh

```diff
@@ -0,0 +1,137 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/experimental/bounding_box.cuh>
+#include <cuspatial/traits.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_vector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/gather.h>
+#include <thrust/scan.h>
+#include <thrust/scatter.h>
+
+namespace cuspatial {
+
+namespace detail {
+
+template <class T,
+          class OffsetIteratorA,
+          class OffsetIteratorB,
+          class VertexIterator,
+          class BoundingBoxIterator>
+BoundingBoxIterator polygon_bounding_boxes(OffsetIteratorA polygon_offsets_first,
+                                           OffsetIteratorA polygon_offsets_last,
+                                           OffsetIteratorB polygon_ring_offsets_first,
+                                           OffsetIteratorB polygon_ring_offsets_last,
+                                           VertexIterator polygon_vertices_first,
+                                           VertexIterator polygon_vertices_last,
+                                           BoundingBoxIterator bounding_boxes_first,
+                                           T expansion_radius,
+                                           rmm::cuda_stream_view stream)
+{
+  auto const num_polygons = std::distance(polygon_offsets_first, polygon_offsets_last);
+  auto const num_rings    = std::distance(polygon_ring_offsets_first, polygon_ring_offsets_last);
+  auto const num_poly_vertices = std::distance(polygon_vertices_first, polygon_vertices_last);
+
+  // Wrapped in an IIFE so `first_ring_offsets` is freed on return
+  auto vertex_ids = [&]() {
+    // TODO: use device_uvector
+    rmm::device_vector<int32_t> vertex_ids(num_poly_vertices);
+    rmm::device_vector<int32_t> first_ring_offsets(num_polygons);
+
+    // Gather the first ring offset for each polygon
+    thrust::gather(rmm::exec_policy(stream),
+                   polygon_offsets_first,
+                   polygon_offsets_last,
+                   polygon_ring_offsets_first,
+                   first_ring_offsets.begin());
+
+    // Scatter the first ring offset into a list of vertex_ids for reduction
+    thrust::scatter(rmm::exec_policy(stream),
+                    thrust::make_counting_iterator(0),
+                    thrust::make_counting_iterator(0) + num_polygons,
+                    first_ring_offsets.begin(),
+                    vertex_ids.begin());
+
+    thrust::inclusive_scan(rmm::exec_policy(stream),
+                           vertex_ids.begin(),
+                           vertex_ids.end(),
+                           vertex_ids.begin(),
+                           thrust::maximum<int32_t>());
+
+    return vertex_ids;
+  }();
```

Comment:
I wonder if you can compute `vertex_ids` on the fly, and will it be faster?
```suggestion
  auto ids_iter = detail::make_counting_transform_iterator(0, []__device__(auto i) {
      auto part_idx = thrust::prev(thrust::upper_bound(thrust::seq, polygon_ring_offsets_first, polygon_ring_offsets_last, i));
      auto geom_idx = thrust::prev(thrust::upper_bound(thrust::seq, polygon_offsets_first, polygon_offsets_last, i));
      return geom_idx;
  }
```
This takes less memory. And for each neighboring `point_idx`, it's guarenteed to find either the same `geometry_idx` or the next one. So the cache hit rate will be high.

Reply:
Great idea. This might change my whole approach.

Reply:
Done.